protocol/proto/base_decoder.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/m3db/m3x/pool"
+)
+
+type decoder struct {
+	r          *bufio.Reader
+	sizeBuffer []byte
+	dataBuffer []byte
+	bytesPool  pool.BytesPool
+}
+
+// NewDecoder decodes a new decoder.
+func NewDecoder(r io.Reader, opts BaseOptions) Decoder {
+	return newDecoder(r, opts)
+}
+
+func newDecoder(r io.Reader, opts BaseOptions) *decoder {
```

Comment:
Cause I did not expose resetReader on the Decoder interface and is only available on *decoder and will be used by the EncodeDecoder, I want user to only reset both writer and reader through EncodeDecoder rather than allow them to reset separately, but I'm open to it.